Add support for Databricks-backed RestStore

aarondav opened the pull request:
## What changes are proposed in this pull request?

This PR adds support for an mlflow tracking URI which refers to a Databricks profile (per the Databricks CLI spec). The URI format proposed is the word "databricks" to refer to the default profile, or "databricks://profile-name" to refer to a different profile.

This PR also changes the interface of how we parse the Databricks CLI to be leak less information between the parsing of the config and the actual HTTP request part of the library. This was done is enable us to parse new configurations over time without having to update every intermediate user (e.g., `secure_verify`).

## How is this patch tested?

- [x] Unit tests - New test suite confirms that the config parsing works as expected, and can invoke the http_request method.
- [x] Manual tests - Tested against a Databricks API endpoint.

aarondav committed:
[WIP] Support Databricks-backed REST API store

aarondav committed:
test

aarondav committed:
Add support for Databricks-backed RestStore

aarondav committed:
lint

aarondav commented on mlflow/experiments.py:
```diff
@@ -16,31 +16,23 @@ def commands():
-              help="Local or S3 URI to store artifacts in (default: inside file store).")
 @click.argument("experiment_name")
-def create(file_store, artifact_root, experiment_name):
+def create(experiment_name):
```
Fixed these methods not to rely on the local file store. @tomasatdatabricks

aarondav commented on mlflow/projects/databricks.py:
```diff
@@ -198,18 +198,13 @@ def _create_databricks_run(tracking_uri, experiment_id, source_name, source_vers
-                                    source_version=source_version,
-                                    entry_point_name=entry_point_name,
-                                    source_type=SourceType.PROJECT)
+    return tracking._create_run(experiment_id=experiment_id,
```
Opted to unconditionally `create_run` and let the warning suffice. In theory, an NFS-mounted local file URI may still be able to work, so we shouldn't completely discount it.

mateiz replied:
I think users might be annoyed if this run still starts and wastes resources, but we can try it.

aarondav replied:
Sorry, wastes resources? This is only related to tracking, i.e., whether we go ahead and create the run object now or later.

aarondav commented on mlflow/store/rest_store.py:
```diff
@@ -38,27 +39,37 @@ def _api_method_to_info():
+        js_dict = http_request(endpoint=endpoint, method=method,
+                               req_body_json=json_body, params=None, **self.http_request_params)
+
+        if 'error_code' in js_dict:
```
Added rudimentary error_code handling here. We should rewrite this section of the code some time; this was just to avoid it from being deserialized as the expected Response object.

smurching replied:
Makes sense. Orthogonally: realized we need to fix error-handling in the open-source server (so that it also returns a well-formed JSON error response instead of a 500 internal error for invalid requests). For example try running `mlflow ui` and hitting `http://localhost:5000/api/2.0/preview/mlflow/experiments/get?experiment_id=some_invalid_id`. Will make a GitHub issue to track this

aarondav replied:
Oh, got it, yeah. They will definitely need to match error behaviors of the UI. We should make sure to do this in the unit/integration test @tomasatdatabricks

aarondav commented on mlflow/utils/rest_utils.py:
```diff
@@ -8,56 +8,77 @@
+                    "https://github.com/databricks/databricks-cli." % profile)
+
+
+def get_databricks_http_request_params_or_fail(profile=None):
```
Refactored method to return a blackbox object which can be passed into `requests` directly, to avoid intermediates having to know about things like `auth`, `hostname`, and `headers`.

aarondav commented on mlflow/utils/rest_utils.py:
```diff
@@ -8,56 +8,77 @@
-                    "CLI is properly configured as described at "
-                    "https://github.com/databricks/databricks-cli." % profile)
+        basic_auth_str = ("%s:%s" % (config.username, config.password)).encode("utf-8")
+    elif config.token:
```
Minor behavior change: we now defer to username/password if set. I can flip the order here if anyone cares, but they should be mostly disjoint.

aarondav commented on mlflow/utils/rest_utils.py:
```diff
@@ -8,56 +8,77 @@
+    }
+
+    secure_verify = True
+    if hasattr(config, 'insecure') and config.insecure:
```
Added support for secure_verify -- previously, were always setting it to False!

smurching commented on mlflow/store/rest_store.py:
```diff
@@ -38,27 +39,37 @@ def _api_method_to_info():
     """ Client for a remote tracking server accessed via REST API calls """
 
-    def __init__(self, hostname):
+    def __init__(self, http_request_params):
```
Nit: maybe document `http_request_params` / consider renaming it to `http_request_kwargs`? It's currently a little confusing to distinguish from query params.

smurching commented on tests/utils/test_rest_utils.py:
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python
```
Nit: remove this line?

smurching commented on tests/utils/test_rest_utils.py:
```diff
@@ -0,0 +1,93 @@
+from mlflow.utils import rest_utils
+
+
+def _mock_profile(expected_profile, config):
```
It could be nice to use `mock.patch.object` (e.g. as we do [here](https://github.com/databricks/mlflow/blob/c7633f228b5527603503fba71f9618898d8c9261/tests/projects/test_databricks.py#L60)) to ensure we clean up / reset `provider` after each test, maybe we could do something like:

```
def _mock_profile(expected_profile, config):
    def mock_get_config_for_profile(profile):
            assert profile == expected_profile
            return config
        provider.get_config_for_profile = mock_get_config_for_profile
    return mock.patch.object(provider, "get_config_for_profile", new=mock_get_config_for_profile)


def test_databricks_params_token():
    with _mock_profile(expected_profile="...", config="..."):
        # perform test, at end of with block `provider` will be reset
```

aarondav replied:
It's so complicated in comparison! I understand how/why the current one works, and if tests fail they fail. I would rather not change unless the test becomes more complicated such that a more compelling reason forms.

smurching replied:
Yeah agreed that it's more complicated, fine with leaving it as is - also agree that if overriding `provider.get_config_for_profile` becomes a problem it'll likely manifest via another test (e.g. one that calls `provider.get_config_for_profile` in another suite) failing & we can address it then.

aarondav replied:
Oh, I'm pretty sure these suites are run in different processes! One suite cannot affect another, right?

smurching commented on tests/utils/test_rest_utils.py:
```diff
@@ -0,0 +1,93 @@
+    _mock_profile("DEFAULT", DatabricksConfig("host", "user", "pass", None, insecure=False))
+    response = rest_utils.databricks_api_request('clusters/list', 'PUT',
+                                                 req_body_json={'a': 'b'}, params='x=y')
+    assert response == {'OK': 'woo'}
```
Should we also test that a `RestException` is thrown when we get a response containing an 'error_code' field?

aarondav replied:
Turns out we don't even have a `test_rest_store`, so created one and added this test to it.

smurching reviewed (commented):
Just a couple of small comments, otherwise this LGTM - thanks @aarondav :)

smurching commented:
Also - we'll eventually need to add docs for specifying a Databricks profile / tracking server, but that can come later.

mateiz commented:
You might as well edit the docs in the same PR IMO. I don't think you need a lot of docs.

aarondav committed:
Fix mocking structure & add rest_store test

aarondav committed:
Add docs

aarondav commented:
Added documentation to mlflow.tracking.set_artifact_uri and [this section of the docs](https://mlflow.org/docs/latest/tracking.html#where-runs-get-recorded). Addressed all comments.

smurching committed:
Tiny typo change

smurching reviewed (approved):
LGTM thanks @aarondav! Did you want to add the test_rest_store tests in this PR? Open to including them in a follow-up / later.

aarondav committed:
Add test!

aarondav commented:
@smurching Whoops, forgot to git add the file. Added!

aarondav committed:
python3 compatibility

smurching commented:
Thanks for the new test, LGTM!
